From: Herton Ronaldo Krzesinski <herton.krzesinski@canonical.com>
Date: Tue, 5 Jul 2011 02:32:58 +0000
Subject: [PATCH] block: check for QUEUE_FLAG_DEAD in elv_completed_request
References: bko#38842,bnc#714215
Patch-Mainline: Not yet

As in __elv_next_request, after blk_cleanup_queue()->elevator_exit() is
called, we can't touch the elevator too. elv_completed_request can be
called after blk_cleanup_queue, as shown in one of the traces on
https://bugs.launchpad.net/bugs/793796, where scsi code ends up being
able to trigger this condition.

Signed-off-by: Herton Ronaldo Krzesinski <herton.krzesinski@canonical.com>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 block/elevator.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/block/elevator.c b/block/elevator.c
index b0b38ce..4d4e04b 100644
--- a/block/elevator.c
+++ b/block/elevator.c
@@ -816,6 +816,7 @@ void elv_completed_request(struct request_queue *q, struct request *rq)
 	if (blk_account_rq(rq)) {
 		q->in_flight[rq_is_sync(rq)]--;
 		if ((rq->cmd_flags & REQ_SORTED) &&
+		    !test_bit(QUEUE_FLAG_DEAD, &q->queue_flags) &&
 		    e->ops->elevator_completed_req_fn)
 			e->ops->elevator_completed_req_fn(q, rq);
 	}
-- 
1.6.0.2

